### PR TITLE
echogen: util for manual testing with echo

### PR DESCRIPTION
**Pull request by stevenctl**
This is a small tool I've been using locally for a week or so. It may or may not be useful to others. 

Rather than running a full set of ITs to get a set of echos to test against, I've found it useful to be able to quickly deploy 1 or 2 services, with specific options like `naked` or `headless` to test edge cases with. Using the tool along with [grpcui](https://github.com/fullstorydev/grpcui) or [grpcurl](https://github.com/fullstorydev/grpcurl) has been pretty nice. 

Changes in this PR:
* cleanup several `fakeNamespace` impls around the codebase
* make `echo.Config` readable from YAML (works around unmarshaling `namespace.Instance` interface type)
* the echogen util itself
* move `EchoPorts` and `WorkloadPorts` to `echo/common` to avoid the `integ` tag in some places

Another thing I'd like to build on top of this, is a way to arbitrarily add `echo.Configs` to a suite at runtime to we can re-use individual ITs (assuming they're following [workload agnostic](https://docs.google.com/document/d/19SP49IoeGSCmNH4jobotWqYL3TSwCO1lI7QjTz_iAEI/edit?usp=sharing&resourcekey=0-5yi-FRW7V0k7cJiEBMBS8g) pattern) in specific scenarios, such as multi-VM testing or testing specific VM distros. These would be tests we'd run on real clusters to test things are working, or anything else that we wouldn't want taking up room in CI. 

---

# echogen

`echogen` is a util for generating Kubernetes manifests from echo configurations.

### Installation:

```bash
go install istio.io/pkg/test/framework/components/echo/echogen
```

### Usage:

```text
echogen [opts] config.yaml
```

The config file is YAML containing a list of  [echo.Config](https://github.com/istio/istio/blob/master/pkg/test/framework/components/echo/config.go#L52) objects:

```yaml
- Service: a
  Namespace: echo
- Service: headless
  Namespace: echo
  Headless: true
```

### Options:

`echogen` supports all options from the test framework that would affect Echo deployments such as: `istio.test
.imagePullSecret`, `istio.test.hub` and several others.

In addition to the framework level options:

```
-out <file>: Write output to the specified file
-dir: If specified, each deployment will be written to a separate file, in a directory named by -out. 
```

### Full Example with gRPC UI

1. Make sure to install [gRPC UI](https://github.com/fullstorydev/grpcui) if you haven't already

2. Create an `echogen` config:

```bash
echo '
- Service: a
  Namespace: echo
- Service: b
  Namespace: echo
' > config.yaml
```

3. Run `echogen`:

```bash
echogen -out echo.yaml config.yaml
```

4. Apply the manifest to the cluster

```bash
kubectl apply -f echo.yaml
```

5. Port-forward the gRPC port (default container port is 17070)

```bash
kubectl -n echo port-forward a-v1-fc649d9fc-59rkj 17070
```

6. Start gRPC UI

```bash
grpcui -plaintext localhost:17070
```

7. Because our echo gRPC service enables reflection, you should be able to open your browser
   and get a user interface that shows all of the possible methods and request options.

   Change the "Method name" to `ForwardEcho`, then in "Request Data" set `url` to `grpc://b:7070` then click `Invoke`

8. (Bonus) If you open the "Raw Request (JSON)" tab, you can re-use that for requests via [grpcurl](https://github.com/fullstorydev/grpcurl)
   without constructing JSON by hand:

```bash
grpcurl -plaintext localhost:17070 EchoTestService/ForwardEcho -d '{"url": "grpc://b:7070"}'
```

### CODE REVIEW

File: pkg/test/framework/components/echo/cmd/echogen/README.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+The config file is YAML containing a list of [echo.Config](https://github.com/istio/istio/blob/master/pkg/test/framework/components/echo/config.go#L52) objects:
+
+```yaml
+- Service: a
+  Namespace: echo
+- Service: headless
+  Namespace: echo
+  Headless: true
+```
+
```
Comment (yangminzhu) [inline pkg/test/framework/components/echo/cmd/echogen/README.md:26]: Could we also document a little bit how to send requests manually with the echo app? Some example commands for `grpcui` or `grpcurl` should be really helpful.

**Review by yangminzhu — COMMENTED** [review]
This is great!

File: pkg/test/framework/components/echo/cmd/echogen/README.md (hunk 1 above)
Comment (stevenctl) [reply pkg/test/framework/components/echo/cmd/echogen/README.md:26]: Added a brief tutorial :)

File: a.yaml [hunk 2]
```diff
@@ -0,0 +1,165 @@
+---apiVersion: v1
```
Comment (yangminzhu) [inline a.yaml:1]: Need a newline here.

**Review by yangminzhu — APPROVED** [review]
LGTM, this is a really nice change, thanks!

File: a.yaml (hunk 2 above)
Comment (stevenctl) [reply a.yaml:1]: Oops, this was an old file that shouldn't have been committed. The golden file shows we don't generate this bad YAML.

**Comment by stevenctl** [comment]
cc @nmittler @Monkeyanator  for T&R/networking review

**Review by nmittler — APPROVED** [review]

**Pull request merged**